creator/ingest_runs/genomic_data_loader.py

```diff
@@ -0,0 +1,361 @@
+
+    def _get_genomic_file_kf_ids(self, ingest_cache, genomic_df):
+        """
+        Part of Step 1
+
+        Extract the genomic_file KF IDs from the ingest cache and add the IDs
+        as a new column to the genomic file DataFrame _genomic_df_.
+        Return _genomic_df_
+        """
+        self.logger.info(
+            "Fetching harmonized genomic file KF IDs from ingest cache ..."
+        )
+        cache_dict = {
+            ast.literal_eval(key)['external_id']: value
+            for key, value in ingest_cache["genomic_file"].items()
+        }
+        hash_df = pd.DataFrame(
+            list(cache_dict.items()),
+            columns=[
+                CONCEPT.GENOMIC_FILE.ID, CONCEPT.GENOMIC_FILE.TARGET_SERVICE_ID
+            ],
+        )
+        genomic_df = genomic_df.merge(
+            right=hash_df,
+            on=CONCEPT.GENOMIC_FILE.ID,
+            how='inner',
+        )
+        return genomic_df
+
+    def _get_seq_experiment_genomic_files(self, harmonized=False):
+        """
+        Part of Step 3
+
+        Get related sequencing experiments for a study's
+        genomic files. Returns the DataFrame from step 4.
+
+        1. Get the study's genomic files
+        2. Get the study's sequencing-experiments
+        3. Get the study's sequencing-experiment-genomic-file links
+        4. Extract the KF IDs from step 1, 2, 3
+        5. Join DataFrames from 1, 2, 3
+        """
+        self.logger.info(
+            "Fetching sequencing experiment info for source genomic files ..."
+        )
+
+        def get_kf_id(link):
+            """
+            The links given by the DataService yield strings of the form
+            /entity-type/kf_id. This function will return the kf_id.
+            """
+            return link.split('/')[-1]
+
+        dfs = {}
+        endpoints = {
+            GEN_FILES,
+            SEQ_EXP_GEN_FILES,
+            SEQ_EXPS,
+        }
+
+        # Steps 1-3
+        filter_dict = {}
+        for endpoint in endpoints:
+            if endpoint == GEN_FILES:
+                filter_dict = {"harmonized": harmonized}
+            dfs[endpoint] = utils.get_entities(
+                DATASERVICE_URL, endpoint,
+                self.study_id, filter_dict=filter_dict
+            )
+
+        # Step 4
+        # Source genomic files
+        gdf = dfs[GEN_FILES][["kf_id", "external_id"]].rename(
+            columns={
+                "kf_id": "gf_kf_id",
+                "external_id": CONCEPT.GENOMIC_FILE.SOURCE_FILE
+            }
+        )
+        # Sequencing experiments
+        sedf = dfs[SEQ_EXPS][
+            ["kf_id", "external_id", "_links.sequencing_center"]
+        ].rename(
+            columns={
+                "kf_id": CONCEPT.SEQUENCING.TARGET_SERVICE_ID,
+                "external_id": CONCEPT.SEQUENCING.ID,
+            }
+        )
+        SCTID = CONCEPT.SEQUENCING.CENTER.TARGET_SERVICE_ID
+        sedf[SCTID] = sedf["_links.sequencing_center"].apply(
```

Comment:
Gio says use .map over .apply for single col transformations bc its 10-100x faster